### PR TITLE
Needed for Ratel #104 - Added readOnly and bestEffort options to query

### PR DESCRIPTION
**Pull request by Aelbannan**
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph-js-http/18)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -14,20 +14,23 @@ and understand how to run and work with Dgraph.
...
-  - [Create a transaction](#create-a-transaction)
-  - [Run a mutation](#run-a-mutation)
-  - [Run a query](#run-a-query)
-  - [Commit a transaction](#commit-a-transaction)
-  - [Check request latency](#check-request-latency)
-  - [Debug mode](#debug-mode)
-- [Development](#development)
-  - [Building the source](#building-the-source)
-  - [Running tests](#running-tests)
+- [dgraph-js-http <img src="https://img.shields.io/npm/v/dgraph-js-http.svg?style=flat" alt="npm version"> <img src="https://img.shields.io/travis/dgraph-io/dgraph-js-http/master.svg?style=flat" alt="Build Status"> <img src="https://img.shields.io/coveralls/github/dgraph-io/dgraph-js-http/master.svg?style=flat" alt="Coverage Status">](#dgraph-js-http-img-src%22httpsimgshieldsionpmvdgraph-js-httpsvgstyleflat%22-alt%22npm-version%22-img-src%22httpsimgshieldsiotravisdgraph-iodgraph-js-httpmastersvgstyleflat%22-alt%22build-status%22-img-src%22httpsimgshieldsiocoverallsgithubdgraph-iodgraph-js-httpmastersvgstyleflat%22-alt%22coverage-status%22)
```
Comment (paulftw) [inline README.md]: This shields thing seems to be showing "no repository found" and if it's unrelated to this change it should be done in a separate PR

File: README.md [hunk 2]
```diff
@@ -229,6 +232,15 @@ Number of people named "Alice": 1
 Alice
 ```
 
+Read-only and best effort queries can be called by passing ```options``` to ```Txn.query``` and ```Txn.queryWithVars```. For example:
+
+```js
+const options = { readOnly: true, bestEffort: true };
+const res = await dgraphClient.newTxn().query(query, options);
```
Comment (paulftw) [inline README.md]: dgraph-js uses options on transaction, not on a query inside transaction. That seems logical, can it be done here?

https://github.com/dgraph-io/dgraph-js/commit/65d45150cf338a2126960e71179d6c9c3c6c4d51

File: src/clientStub.ts [hunk 3]
```diff
@@ -109,6 +109,18 @@ export class DgraphClientStub {
               value: "true",
             });
           }
+          if (req.readOnly) {
```
Comment (paulftw) [inline src/clientStub.ts:112]: see comment above. this would become txn options

File: src/txn.ts [hunk 4]
```diff
@@ -52,18 +52,25 @@ export class Txn {
...
+        options: { debug?: boolean, readOnly?: boolean, bestEffort?: boolean } = {},
     ): Promise<Response> {
         if (this.finished) {
             this.dc.debug(`Query request (ERR_FINISHED):\nquery = ${q}\nvars = ${vars}`);
             throw ERR_FINISHED;
         }
 
+        if (options.bestEffort && !options.readOnly) {
+            this.dc.debug('Best effort only works with read-only queries.');
+            throw ERR_FINISHED;
```
Comment (paulftw) [inline src/txn.ts]: doesn't sound like the right ERR object to throw

**Review by paulftw — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (Aelbannan) [reply README.md]: Sure, I can move that to the transaction

File: src/txn.ts (hunk 4 above)
Comment (Aelbannan) [reply src/txn.ts]: Good point

File: README.md (hunk 1 above)
Comment (Aelbannan) [reply README.md]: Not sure what this is, I'll take a look at it

Comment (Aelbannan) [reply README.md]: Removed

File: README.md (hunk 2 above)
Comment (Aelbannan) [reply README.md]: Done

File: src/txn.ts (hunk 4 above)
Comment (Aelbannan) [reply src/txn.ts]: Done

File: src/clientStub.ts (hunk 3 above)
Comment (Aelbannan) [reply src/clientStub.ts:112]: Seems like dgraph-js is doing it like this (storing readOnly/bestEffort in Request). Refer to here:
https://github.com/dgraph-io/dgraph-js/blob/c1ad859ea9f683ca7efb51b481e56ec868a7da1a/src/txn.ts#L79

File: src/types.ts [hunk 5]
```diff
@@ -64,10 +66,17 @@ export interface TxnContext {
     aborted?: boolean | null;
     keys?: string[] | null;
     preds?: string[] | null;
+    readOnly: boolean;
```
Comment (paulftw) [inline src/types.ts:69]: Is this for the server's response? I don't think we're ever reading these fields, are you sure you need RO/BE flags *both* here and in TxnOptions?

**Review by paulftw — APPROVED** [review]
